### ISSUE TITLE
Using the Read and Write method of S7MultiVar prompts ISO: Invalid PDU received

### ISSUE THREAD

**Issue by tzm2270969436** [fbarresi/Sharp7#43]
If my PLC PDU size value is 240, but I need to read or write a length of 500 bytes,25 bytes per address, how should I perform batch processing operations, calculate the byte length sent to the PLC, and determine the byte length of the PLC response? What is the formula for calculating the byte length of sending and receiving messages

```
var client = new S7Client();
int result = client.ConnectTo("127.0.0.1", 0, 2); 
if (result != 0)
{
    Console.WriteLine(client.ErrorText(result));
    return;
}

var items = new List<S7Tag>();
for (int i = 0; i < 20; i++)
{
    items.Add(new S7Tag
    {
        Area = (int)S7Area.DB,
        DBNumber = 1, 
        Start = i * 25,
        Elements = 25, 
        WordLen = (int)S7WordLength.Char,
    });
};
var length = tags.Sum(x => x.WordLen.DataSizeByte() * x.Elements);
var buffer = new byte[length];
var offset = 0;
var multiVar = new S7MultiVar(client);
foreach (var tag in tags)
{
     multiVar.Add(tag, ref buffer, offset);
     offset += tag.WordLen.DataSizeByte() * tag.Elements;
}
var readResult = multiVar.Read();
if (readResult != 0)
{
    Console.WriteLine(client.ErrorText(readResult));
    return;
}
```

**Comment by hebelmx** [comment]
you are almost there, don't send more than 16 elements, play safe and don't send more than 15!

**Comment by MohamedGobran** [comment]
Looking at your code, I can see exactly what's happening here. You're hitting PDU limits because you're trying to read 500 bytes across 20 MultiVar items, but your PLC only supports 240-byte PDU.

## S7 protocol PDU Math

The S7 protocol has specific overhead for MultiVar operations:
- **Request PDU**: `(Number of items + 1) × 12 bytes`
- **Response PDU**: `14 + (4 × Number of items) + actual data`

With your 20 items:
- Request: (20 + 1) × 12 = **252 bytes**  (over your 240 limit)  
- Response: 14 + (4 × 20) + 500 = **594 bytes**  (way over)

That's exactly why you're getting PDU errors.

## Solution 1: MultiVar Batching

**Important**: Sharp7 does **NOT** automatically split MultiVar requests into multiple operations. You must manually batch them to stay within PDU limits.

If you want to stick with MultiVar, you need batching. Here's the formula:

```csharp
// Calculate max items per batch (response is usually the bottleneck)
// 14 + (4 × items) + (25 × items) = 14 + 29 × items <= 240
int maxItems = (240 - 14) / 29; // = 7 items max per batch
```

```csharp
var client = new S7Client();
int result = client.ConnectTo("127.0.0.1", 0, 2); 
if (result != 0)
{
    Console.WriteLine(client.ErrorText(result));
    return;
}

// Create all your items
var allItems = new List<S7Tag>();
for (int i = 0; i < 20; i++)
{
    allItems.Add(new S7Tag
    {
        Area = (int)S7Area.DB,
        DBNumber = 1, 
        Start = i * 25,
        Elements = 25, 
        WordLen = (int)S7WordLength.Char,
    });
}

var buffer = new byte[500];
int bufferOffset = 0;

// Process in batches of 7 items
for (int i = 0; i < allItems.Count; i += 7)
{
    var batchSize = Math.Min(7, allItems.Count - i);
    var multiVar = new S7MultiVar(client);
    
    for (int j = 0; j < batchSize; j++)
    {
        var item = allItems[i + j];
        multiVar.Add(item, ref buffer, bufferOffset);
        bufferOffset += 25;
    }
    
    var readResult = multiVar.Read();
    if (readResult != 0)
    {
        Console.WriteLine($"Batch {i/7 + 1} failed: {client.ErrorText(readResult)}");
        return;
    }
}
```

You'll need **3 batches** (7 + 7 + 6 items). 

## Solution 2: Block Read (Better for Your Case)

But honestly? Since you're reading **sequential data from the same DB** (addresses 0, 25, 50, 75...), MultiVar is overkill. Just read the whole block:

```csharp
var client = new S7Client();
int result = client.ConnectTo("127.0.0.1", 0, 2); 
if (result != 0)
{
    Console.WriteLine(client.ErrorText(result));
    return;
}

// Read all 500 bytes as one continuous block
byte[] buffer = new byte[500];
int readResult = client.DBRead(1, 0, 500, buffer);

if (readResult == 0)
{
    // Extract your 25-byte chunks
    for (int i = 0; i < 20; i++)
    {
        // Your data for item i starts at buffer[i * 25]
        // and goes for 25 bytes
    }
}
else
{
    Console.WriteLine($"Read failed: {client.ErrorText(readResult)}");
}
```

**Why this is better:**
- Much simpler code
- **Sharp7 handles PDU splitting automatically for block reads** (unlike MultiVar which requires manual batching)
- Less protocol overhead
- Perfect for sequential data like yours



## Bottom Line

**MultiVar** = Great for scattered data across different areas, but **requires manual batching** for PDU management  
**Block Read** = Perfect for continuous data in the same area + **automatic PDU handling**

Try the block read first - it's designed exactly for what you're doing and Sharp7 will automatically handle any PDU splitting needed.

**Comment by MOP3E** [comment]
My universal choice for sorting variables. It takes some time at the start to sort the variables, but you do not need to take into account the current size of the PLC PDU.

**Variables**
int i  - the index.
List<List<Tag*>> pduGroups - list of tags sorted by PDU.
Tag* is just a kind of structure for storing descriptions of tags associated with variables in the PLC (name, type of PLC variable, address in the PLC, etc.).

**Initial state**
1. The pduGroups list contains one element, a list that lists all the tags that need to be read.
2. Index i = 0.

**The reading cycle**
1. If i >= pduGroups.Count, finish the job.
2. Try to read the tags of the item with the index i.
3. If everything is OK, go to step 7.
4. If the reading error is not a PDU error, shut down with an error.
5. Move the last tag from pduGroups[i] to the beginning of pduGroups[i+1] (if pduGroups[i+1] not exists, create it)
6. Go to step 2.
7. Increase index i by 1 and go to step 1.
8. End.
